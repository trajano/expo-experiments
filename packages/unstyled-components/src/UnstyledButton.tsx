import { Children, FC, isValidElement, ReactElement } from 'react';
import {
  StyleProp,
  StyleSheet,
  View,
  ViewProps,
  ViewStyle,
  I18nManager,
  Animated,

} from 'react-native';

/**
 * Props for the UnstyledButton component.
 * Extends from ViewProps to include all standard View properties.
 *
 * @property {StyleProp<ViewStyle>} contentContainerStyle - Optional style for the main content container.
 */
type UnstyledButtonProps = ViewProps & {
  contentContainerStyle?: StyleProp<ViewStyle>;
};

/**
 * UnstyledButton Component
 *
 * This component serves as a customizable button that provides "slots" for a start icon, end icon,
 * and main content. It respects RTL (Right-To-Left) layout settings automatically using `I18nManager`.
 *
 * @example
 * <UnstyledButton testID="test-id" style={{ padding: 10 }}>
 *   <UnstyledButton.StartIcon>
 *     <Image source={{ uri: 'startImage' }} style={{ width: 20, height: 20 }} />
 *   </UnstyledButton.StartIcon>
 *   <Text>Hello world</Text>
 *   <Text>Stinky Tofu</Text>
 *   <UnstyledButton.EndIcon>
 *     <Text>âœ“</Text>
 *   </UnstyledButton.EndIcon>
 * </UnstyledButton>
 */
export const UnstyledButton: FC<UnstyledButtonProps> & {
  StartIcon: FC<ViewProps>;
  EndIcon: FC<ViewProps>;
} = ({ style, contentContainerStyle, children, testID, ...rest }) => {
  const startIcon = Children.toArray(children).find(
    (child) => isValidElement(child) && child.type === UnstyledButton.StartIcon,
  );

  const endIcon = Children.toArray(children).find(
    (child) => isValidElement(child) && child.type === UnstyledButton.EndIcon,
  );

  const mainContent = Children.toArray(children).filter(
    (child) => !(isValidElement(child) && (child.type === UnstyledButton.StartIcon || child.type === UnstyledButton.EndIcon)),
  );

  return (
    <View
      testID={testID}
      style={style ? [styles.buttonStyle, style] : styles.buttonStyle}
      {...rest}
    >
      {startIcon}
      <View style={contentContainerStyle}>
        {mainContent}
      </View>
      {endIcon}

    </View>
  );
};

/**
 * StartIcon Component
 *
 * A slot for placing an icon at the start of the UnstyledButton. Can be styled and customized
 * as needed. Typically used for leading icons.
 */
const StartIcon: FC<ViewProps> = ({ children, ...props }) => (
  <View {...props}>{children}</View>
);

/**
 * EndIcon Component
 *
 * A slot for placing an icon at the end of the UnstyledButton. Can be styled and customized
 * as needed. Typically used for trailing icons.
 */
const EndIcon: FC<ViewProps> = ({ children, ...props }) => (
  <View {...props}>{children}</View>
);

// Assign display names for better debugging and visibility in React DevTools
StartIcon.displayName = 'UnstyledButton.StartIcon';
EndIcon.displayName = 'UnstyledButton.EndIcon';

// Attach StartIcon and EndIcon as static properties to the main component
UnstyledButton.StartIcon = StartIcon;
UnstyledButton.EndIcon = EndIcon;

// Styles for the UnstyledButton component
const styles = StyleSheet.create({
  buttonStyle: {
    flexDirection: I18nManager.isRTL ? 'row-reverse' : 'row', // Adjust layout direction based on RTL setting
    alignItems: 'center', // Vertically center all children
    justifyContent: 'center', // Horizontally center if space available
  },
});

export const AnimatedUnstyledButton = Animated.createAnimatedComponent(UnstyledButton);