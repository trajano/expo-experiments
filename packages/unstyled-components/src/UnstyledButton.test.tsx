import { render } from '@testing-library/react-native';
import { Text, Image } from 'react-native';
import { UnstyledButton } from './UnstyledButton'; // Adjust the import path as necessary

describe('UnstyledButton Component', () => {
  it('should render the wrapper, start icon, main content, and end icon correctly', () => {
    const { getByTestId, getByText, toJSON } = render(
      <UnstyledButton
        style={{ padding: 10, flexDirection: 'row', alignItems: 'center' }}
        contentContainerStyle={{ justifyContent: 'center' }}
        testID="test-id"
      >
        <UnstyledButton.StartIcon style={{ marginRight: 5 }}>
          <Image source={{ uri: 'startImage' }} style={{ width: 20, height: 20 }} />
        </UnstyledButton.StartIcon>
        <Text>Hello world</Text>
        <Text>Stinky Tofu</Text>
        <UnstyledButton.EndIcon style={{ marginLeft: 5 }}>
          <Text>âœ“</Text>
        </UnstyledButton.EndIcon>
      </UnstyledButton>
    );

    // Check for wrapper
    const wrapper = getByTestId('test-id_wrapper');
    expect(wrapper).toBeTruthy();

    // Check for start icon
    const startIcon = getByTestId('test-id_startIcon');
    expect(startIcon).toBeTruthy();

    // Check for main content (Text elements)
    const mainContent = getByTestId('test-id');
    expect(mainContent).toBeTruthy();

    // Verify presence of specific texts
    expect(getByText('Hello world')).toBeTruthy();
    expect(getByText('Stinky Tofu')).toBeTruthy();

    // Check for end icon
    const endIcon = getByTestId('test-id_endIcon');
    expect(endIcon).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render without start and end icons if they are not provided', () => {
    const { getByTestId, queryByTestId, getByText, toJSON } = render(
      <UnstyledButton
        testID="test-id"
      >
        <Text>Hello world</Text>
      </UnstyledButton>
    );

    // Check for wrapper
    const wrapper = getByTestId('test-id_wrapper');
    expect(wrapper).toBeTruthy();

    // Check for main content (Text elements)
    const mainContent = getByTestId('test-id');
    expect(mainContent).toBeTruthy();

    // Verify presence of specific texts
    expect(getByText('Hello world')).toBeTruthy();

    // Check that start and end icons are not rendered
    expect(queryByTestId('test-id_startIcon')).toBeNull();
    expect(queryByTestId('test-id_endIcon')).toBeNull();
    expect(toJSON()).toMatchSnapshot();
  });
});
