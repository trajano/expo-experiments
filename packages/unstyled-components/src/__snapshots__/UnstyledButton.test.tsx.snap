// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnstyledButton Component should render the wrapper, start icon, main content, and end icon correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
      {
        "alignItems": "center",
        "flexDirection": "row",
        "padding": 10,
      },
    ]
  }
  testID="test-id_wrapper"
>
  <View
    style={
      {
        "marginRight": 5,
      }
    }
    testID="test-id_startIcon"
  >
    <Image
      source={
        {
          "uri": "startImage",
        }
      }
      style={
        {
          "height": 20,
          "width": 20,
        }
      }
    />
  </View>
  <View
    style={
      {
        "justifyContent": "center",
      }
    }
    testID="test-id"
  >
    <View
      style={
        {
          "marginRight": 5,
        }
      }
    >
      <Image
        source={
          {
            "uri": "startImage",
          }
        }
        style={
          {
            "height": 20,
            "width": 20,
          }
        }
      />
    </View>
    <Text>
      Hello world
    </Text>
    <Text>
      Stinky Tofu
    </Text>
    <View
      style={
        {
          "marginLeft": 5,
        }
      }
    >
      <Text>
        ✓
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "marginLeft": 5,
      }
    }
    testID="test-id_endIcon"
  >
    <Text>
      ✓
    </Text>
  </View>
</View>
`;

exports[`UnstyledButton Component should render without start and end icons if they are not provided 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
  testID="test-id_wrapper"
>
  <View
    testID="test-id"
  >
    <Text>
      Hello world
    </Text>
  </View>
</View>
`;
